internal/app/project/gradle.go

```diff
@@ -1,13 +1,205 @@
 package project
 
-func GuessGradleTasks(projectDir string) (runTask, buildTask string, err error) {
-	//	If the build gradle has Ktor plugin and it's applied
-	//		Use run and classes command
-	//	Else
-	//		Search one level in the subdirectories for build.gradle.kts
-	//		For each found
-	//			If the build gradle has Ktor plugin and it's applied
-	//				Use :subdir:run and :subdir:classes command
+import (
+	"fmt"
+	"github.com/ktorio/ktor-cli/internal/app/lang/gradle"
+	"github.com/ktorio/ktor-cli/internal/app/lang/toml"
+	"github.com/ktorio/ktor-cli/internal/app/utils"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"unicode"
+)
 
+type semver struct {
+	valid               bool
+	major, minor, patch int
+}
+
+var minPluginVersion = &semver{major: 3, minor: 0, patch: 3} // Subject to change
+
+func GuessGradleTasks(projectDir string) (runTask, buildTask string, guessed bool) {
+	buildPath := filepath.Join(projectDir, "build.gradle.kts")
+	if !utils.Exists(buildPath) {
+		return
+	}
+
+	if pVersion, found := searchKtorPlugin(projectDir); found && pVersion.valid && pVersion.aboveOrEqual(minPluginVersion) {
+		runTask = "run"
+		buildTask = "classes"
+		guessed = true
+	} else {
+		entries, err := os.ReadDir(projectDir)
+		if err != nil {
+			return
+		}
+
+		var candidateDirs []string
+		for _, e := range entries {
+			if e.IsDir() && !strings.HasPrefix(e.Name(), ".") && e.Name() != "gradle" {
+				candidateDirs = append(candidateDirs, e.Name())
+			}
+		}
+
+		// TODO: Check dirs concurrently
+		for _, dirName := range candidateDirs {
+			candidateDir := filepath.Join(projectDir, dirName)
+			if utils.Exists(filepath.Join(candidateDir, "build.gradle.kts")) {
+				if pVersion, found = searchKtorPlugin(candidateDir); found && pVersion.valid && pVersion.aboveOrEqual(minPluginVersion) {
+					runTask = fmt.Sprintf(":%s:run", dirName)
+					buildTask = fmt.Sprintf(":%s:classes", dirName)
+					guessed = true
+					break
+				}
+			}
+		}
+	}
+
+	return
+}
+func searchKtorPlugin(dir string) (version semver, found bool) {
+	buildPath := filepath.Join(dir, "build.gradle.kts")
+	buildRoot, err, _ := gradle.ParseBuildFile(buildPath)
+
+	if err != nil {
+		return
+	}
+
+	found = true
+	var catalogKeys []string
+	for _, p := range buildRoot.Plugins.List {
+		if p.IsCatalog && p.Applied {
+			if segments := strings.Split(p.Id, "."); len(segments) > 0 {
+				off := len(segments) - 1
+
+				for ; off >= 0; off-- {
+					if segments[off] == "plugins" {
+						off++
+						break
+					}
+				}
+
+				catalogKeys = append(catalogKeys, strings.Join(segments[off:], "-"))
+			}
+		}
+
+		if p.Prefix == "id" && p.Id == "io.ktor.plugin" && p.Applied {
+			if p.Version != "" {
+				version = parseVersion(p.Version)
+			}
+			return
+		}
+	}
+
+	if len(catalogKeys) > 0 {
+		hasCatalog := false
+		catalogName := "libs.versions.toml"
```

Comment:
Minor. Should this work for catalogs with non-default name? Something different from libs

Reply:
This and the `add` command currently don't work with non-default filenames. We could support this use case if someone requested it.